### PR TITLE
feat(gatsby-plugin-sharp): run a single transformation for lazy images

### PR DESCRIPTION
**Pull request by vladar**
## Description

Before this PR lazy images run all possible image transformations when a single transformed image is requested. After this PR only the requested image variant is processed and returned.

All the remaining transformations are still stored in the cached job, so whenever another image variant is requested - it will be also lazily transformed and returned.

Tried it on my current image-heavy test site. All transforms:
```
success Generating image thumbnails - 53.272s - 72/72 1.35/s
```

The same example with this PR:
```
success Generating image thumbnails - 25.650s - 24/24 0.94/s
```

## Related Issues

#27603 - umbrella discussion

### CODE REVIEW

File: packages/gatsby-plugin-sharp/src/gatsby-node.js [hunk 1]
```diff
@@ -78,15 +78,58 @@ exports.onCreateDevServer = async ({ app, cache, reporter }) => {
...
 }
 
+// Split the job into two jobs:
+//  - first job with a single operation matching requestedPathOnDisk
+//  - second job with all other operations
+// so the two resulting jobs are only different by their operations
+function splitOperationsByRequestedFile(job, requestedPathOnDisk) {
+  const matchingJob = {
+    ...job,
+    args: { ...job, operations: [] },
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/gatsby-node.js:110]: I suppose this is job.args?
```suggestion
    args: { ...job.args, operations: [] },
```

File: packages/gatsby-plugin-sharp/src/gatsby-node.js [hunk 2]
```diff
@@ -78,15 +78,58 @@ exports.onCreateDevServer = async ({ app, cache, reporter }) => {
...
+  }
+  const jobWithRemainingOperations = {
+    ...job,
+    args: { ...job, operations: [] },
+  }
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/gatsby-node.js:115]: I suppose this is job.args?
```suggestion
    args: { ...job.args, operations: [] },
```

File: packages/gatsby-plugin-sharp/src/gatsby-node.js [hunk 3]
```diff
@@ -78,15 +78,58 @@ exports.onCreateDevServer = async ({ app, cache, reporter }) => {
...
+      // There are still some operations pending for this job - replace the cached job
+      await cache.cache.set(jobContentDigest, jobWithRemainingOperations)
+    } else {
+      // No operations left to process - purge the cache
+      await cache.cache.del(jobContentDigest)
+    }
+
     return res.sendFile(pathOnDisk)
   })
 }
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/gatsby-node.js:101]: Could you add some more context why this is necessary? (I assume it's to do less work so HTTP 6 requests limit doesn't take too long)

**Review by wardpeet — APPROVED** [review]
Looking good, some small comments.

File: packages/gatsby-plugin-sharp/src/gatsby-node.js (hunk 3 above)
Comment (vladar) [reply packages/gatsby-plugin-sharp/src/gatsby-node.js:101]: Yup, just to transform only what's needed by the browser. This makes lazy images to load least 2+ times faster and so free browser queue earler. I'll add a comment.

**Review by wardpeet — APPROVED** [review]

**Comment by vladar** [comment]
Windows unit tests seem to be failing for unrelated reasons (also fail on other PRs), so merging as-is.

**Pull request merged**